gsql stuck if connecting to ShardingSphere-Proxy openGauss and specified a non-exist database name

## Bug Report

### Which version of ShardingSphere did you use?

c3596089f4bfccd5d2b02b164c203f72e5d779be

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?

ShardingSphere-Proxy openGauss

### Expected behavior

gsql client got error message immediately.

```
ERROR:  FATAL: database "nonexistdb" does not exist
```

### Actual behavior

gsql client stuck.

```
[ERROR] 2022-09-24 11:19:45.610 [epollEventLoopGroup-3-2] o.a.s.p.f.n.FrontendChannelInboundHandler - Exception occur: 
org.apache.shardingsphere.dialect.exception.syntax.database.UnknownDatabaseException: null
	at org.apache.shardingsphere.proxy.frontend.opengauss.authentication.OpenGaussAuthenticationHandler.loginWithSCRAMSha256Password(OpenGaussAuthenticationHandler.java:80)
	at org.apache.shardingsphere.proxy.frontend.opengauss.authentication.OpenGaussAuthenticationEngine.processPasswordMessage(OpenGaussAuthenticationEngine.java:115)
	at org.apache.shardingsphere.proxy.frontend.opengauss.authentication.OpenGaussAuthenticationEngine.authenticate(OpenGaussAuthenticationEngine.java:92)
	at org.apache.shardingsphere.proxy.frontend.netty.FrontendChannelInboundHandler.authenticate(FrontendChannelInboundHandler.java:77)
	at org.apache.shardingsphere.proxy.frontend.netty.FrontendChannelInboundHandler.channelRead(FrontendChannelInboundHandler.java:69)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:336)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:308)
	at io.netty.handler.codec.ByteToMessageCodec.channelRead(ByteToMessageCodec.java:103)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:357)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
	at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:800)
	at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:499)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:397)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)



Exception in thread "ShardingSphere-Command-0" java.lang.NullPointerException: Cannot invoke "Object.hashCode()" because "key" is null
	at java.base/java.util.concurrent.ConcurrentHashMap.replaceNode(ConcurrentHashMap.java:1111)
	at java.base/java.util.concurrent.ConcurrentHashMap.remove(ConcurrentHashMap.java:1102)
	at org.apache.shardingsphere.mode.process.ShowProcessListManager.removeProcessStatement(ShowProcessListManager.java:116)
	at org.apache.shardingsphere.mode.process.GovernanceExecuteProcessReporter.reportRemove(GovernanceExecuteProcessReporter.java:78)
	at org.apache.shardingsphere.infra.executor.sql.process.ExecuteProcessEngine.lambda$finishConnection$1(ExecuteProcessEngine.java:73)
	at java.base/java.util.Optional.ifPresent(Optional.java:178)
	at org.apache.shardingsphere.infra.executor.sql.process.ExecuteProcessEngine.finishConnection(ExecuteProcessEngine.java:73)
	at org.apache.shardingsphere.proxy.frontend.netty.FrontendChannelInboundHandler.closeAllResources(FrontendChannelInboundHandler.java:108)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)

```


### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.

```
gsql -h 127.0.0.1 -p 55433 -U gaussdb nonexistdb -WPassword
```

--- a/db-protocol/opengauss/src/main/java/org/apache/shardingsphere/db/protocol/opengauss/constant/OpenGaussProtocolVersion.java
+++ b/db-protocol/opengauss/src/main/java/org/apache/shardingsphere/db/protocol/opengauss/constant/OpenGaussProtocolVersion.java
@@ -1,0 +1,35 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.shardingsphere.db.protocol.opengauss.constant;
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
+/**
+ * Protocol version of openGauss.
+ */
+@RequiredArgsConstructor
+@Getter
+public enum OpenGaussProtocolVersion {
+    
+    PROTOCOL_350(3 << 16 | 50),
+    
+    PROTOCOL_351(3 << 16 | 51);
+    
+    private final int version;
+}

--- a/db-protocol/opengauss/src/main/java/org/apache/shardingsphere/db/protocol/opengauss/packet/authentication/OpenGaussAuthenticationSCRAMSha256Packet.java
+++ b/db-protocol/opengauss/src/main/java/org/apache/shardingsphere/db/protocol/opengauss/packet/authentication/OpenGaussAuthenticationSCRAMSha256Packet.java
@@ -18,6 +18,7 @@
 package org.apache.shardingsphere.db.protocol.opengauss.packet.authentication;
 
 import lombok.RequiredArgsConstructor;
+import org.apache.shardingsphere.db.protocol.opengauss.constant.OpenGaussProtocolVersion;
 import org.apache.shardingsphere.db.protocol.postgresql.packet.identifier.PostgreSQLIdentifierPacket;
 import org.apache.shardingsphere.db.protocol.postgresql.packet.identifier.PostgreSQLIdentifierTag;
 import org.apache.shardingsphere.db.protocol.postgresql.packet.identifier.PostgreSQLMessagePacketType;
@@ -33,9 +34,13 @@ public final class OpenGaussAuthenticationSCRAMSha256Packet implements PostgreSQ
     
     private static final int PASSWORD_STORED_METHOD_SHA256 = 2;
     
+    private final int version;
+    
     private final byte[] random64Code;
     
     private final byte[] token;
+    
+    private final byte[] serverSignature;
     
     private final int serverIteration;
     
@@ -45,7 +50,12 @@ public final class OpenGaussAuthenticationSCRAMSha256Packet implements PostgreSQ
         payload.writeInt4(PASSWORD_STORED_METHOD_SHA256);
         payload.writeBytes(random64Code);
         payload.writeBytes(token);
-        payload.writeInt4(serverIteration);
+        if (version < OpenGaussProtocolVersion.PROTOCOL_350.getVersion()) {
+            payload.writeBytes(serverSignature);
+        }
+        if (OpenGaussProtocolVersion.PROTOCOL_351.getVersion() == version) {
+            payload.writeInt4(serverIteration);
+        }
     }
     
     @Override

--- a/db-protocol/opengauss/src/test/java/org/apache/shardingsphere/db/protocol/opengauss/packet/authentication/OpenGaussAuthenticationSCRAMSha256PacketTest.java
+++ b/db-protocol/opengauss/src/test/java/org/apache/shardingsphere/db/protocol/opengauss/packet/authentication/OpenGaussAuthenticationSCRAMSha256PacketTest.java
@@ -17,6 +17,7 @@
 
 package org.apache.shardingsphere.db.protocol.opengauss.packet.authentication;
 
+import org.apache.shardingsphere.db.protocol.opengauss.constant.OpenGaussProtocolVersion;
 import org.apache.shardingsphere.db.protocol.postgresql.packet.identifier.PostgreSQLMessagePacketType;
 import org.apache.shardingsphere.db.protocol.postgresql.payload.PostgreSQLPacketPayload;
 import org.junit.Test;
@@ -32,23 +33,50 @@ public final class OpenGaussAuthenticationSCRAMSha256PacketTest {
     
     private static final byte[] TOKEN = new byte[8];
     
-    private static final int SERVER_ITERATION = 2048;
-    
-    private final OpenGaussAuthenticationSCRAMSha256Packet packet = new OpenGaussAuthenticationSCRAMSha256Packet(RANDOM_64_CODE, TOKEN, SERVER_ITERATION);
+    private static final byte[] SERVER_SIGNATURE = new byte[64];
     
     @Test
-    public void assertWrite() {
+    public void assertWriteProtocol300Packet() {
         PostgreSQLPacketPayload payload = mock(PostgreSQLPacketPayload.class);
+        OpenGaussAuthenticationSCRAMSha256Packet packet = new OpenGaussAuthenticationSCRAMSha256Packet(
+                OpenGaussProtocolVersion.PROTOCOL_350.getVersion(), RANDOM_64_CODE, TOKEN, SERVER_SIGNATURE, 2048);
         packet.write(payload);
         verify(payload).writeInt4(10);
         verify(payload).writeInt4(2);
         verify(payload).writeBytes(RANDOM_64_CODE);
         verify(payload).writeBytes(TOKEN);
-        verify(payload).writeInt4(SERVER_ITERATION);
+        verify(payload).writeBytes(SERVER_SIGNATURE);
+    }
+    
+    @Test
+    public void assertWriteProtocol350Packet() {
+        PostgreSQLPacketPayload payload = mock(PostgreSQLPacketPayload.class);
+        OpenGaussAuthenticationSCRAMSha256Packet packet = new OpenGaussAuthenticationSCRAMSha256Packet(
+                OpenGaussProtocolVersion.PROTOCOL_350.getVersion(), RANDOM_64_CODE, TOKEN, SERVER_SIGNATURE, 2048);
+        packet.write(payload);
+        verify(payload).writeInt4(10);
+        verify(payload).writeInt4(2);
+        verify(payload).writeBytes(RANDOM_64_CODE);
+        verify(payload).writeBytes(TOKEN);
+    }
+    
+    @Test
+    public void assertWriteProtocol351Packet() {
+        PostgreSQLPacketPayload payload = mock(PostgreSQLPacketPayload.class);
+        OpenGaussAuthenticationSCRAMSha256Packet packet = new OpenGaussAuthenticationSCRAMSha256Packet(
+                OpenGaussProtocolVersion.PROTOCOL_351.getVersion(), RANDOM_64_CODE, TOKEN, SERVER_SIGNATURE, 10000);
+        packet.write(payload);
+        verify(payload).writeInt4(10);
+        verify(payload).writeInt4(2);
+        verify(payload).writeBytes(RANDOM_64_CODE);
+        verify(payload).writeBytes(TOKEN);
+        verify(payload).writeInt4(10000);
     }
     
     @Test
     public void assertIdentifierTag() {
+        OpenGaussAuthenticationSCRAMSha256Packet packet = new OpenGaussAuthenticationSCRAMSha256Packet(
+                OpenGaussProtocolVersion.PROTOCOL_351.getVersion(), RANDOM_64_CODE, TOKEN, SERVER_SIGNATURE, 10000);
         assertThat(packet.getIdentifier(), is(PostgreSQLMessagePacketType.AUTHENTICATION_REQUEST));
     }
 }

--- a/db-protocol/postgresql/src/main/java/org/apache/shardingsphere/db/protocol/postgresql/packet/handshake/PostgreSQLComStartupPacket.java
+++ b/db-protocol/postgresql/src/main/java/org/apache/shardingsphere/db/protocol/postgresql/packet/handshake/PostgreSQLComStartupPacket.java
@@ -17,6 +17,7 @@
 
 package org.apache.shardingsphere.db.protocol.postgresql.packet.handshake;
 
+import lombok.Getter;
 import org.apache.shardingsphere.db.protocol.postgresql.packet.PostgreSQLPacket;
 import org.apache.shardingsphere.db.protocol.postgresql.payload.PostgreSQLPacketPayload;
 
@@ -34,10 +35,14 @@ public final class PostgreSQLComStartupPacket implements PostgreSQLPacket {
     
     private static final String CLIENT_ENCODING_KEY = "client_encoding";
     
+    @Getter
+    private final int version;
+    
     private final Map<String, String> parametersMap = new HashMap<>();
     
     public PostgreSQLComStartupPacket(final PostgreSQLPacketPayload payload) {
-        payload.skipReserved(8);
+        payload.skipReserved(4);
+        version = payload.readInt4();
         while (payload.bytesBeforeZero() > 0) {
             parametersMap.put(payload.readStringNul(), payload.readStringNul());
         }

--- a/proxy/frontend/opengauss/src/main/java/org/apache/shardingsphere/proxy/frontend/opengauss/authentication/OpenGaussAuthenticationEngine.java
+++ b/proxy/frontend/opengauss/src/main/java/org/apache/shardingsphere/proxy/frontend/opengauss/authentication/OpenGaussAuthenticationEngine.java
@@ -19,7 +19,9 @@ package org.apache.shardingsphere.proxy.frontend.opengauss.authentication;
 
 import com.google.common.base.Strings;
 import io.netty.channel.ChannelHandlerContext;
+import org.apache.shardingsphere.authority.rule.AuthorityRule;
 import org.apache.shardingsphere.db.protocol.CommonConstants;
+import org.apache.shardingsphere.db.protocol.opengauss.constant.OpenGaussProtocolVersion;
 import org.apache.shardingsphere.db.protocol.opengauss.packet.authentication.OpenGaussAuthenticationSCRAMSha256Packet;
 import org.apache.shardingsphere.db.protocol.payload.PacketPayload;
 import org.apache.shardingsphere.db.protocol.postgresql.constant.PostgreSQLServerInfo;
@@ -33,6 +35,8 @@ import org.apache.shardingsphere.db.protocol.postgresql.packet.identifier.Postgr
 import org.apache.shardingsphere.db.protocol.postgresql.payload.PostgreSQLPacketPayload;
 import org.apache.shardingsphere.dialect.postgresql.exception.authority.EmptyUsernameException;
 import org.apache.shardingsphere.dialect.postgresql.exception.protocol.ProtocolViolationException;
+import org.apache.shardingsphere.infra.metadata.user.Grantee;
+import org.apache.shardingsphere.proxy.backend.context.ProxyContext;
 import org.apache.shardingsphere.proxy.backend.handler.admin.postgresql.PostgreSQLCharacterSets;
 import org.apache.shardingsphere.proxy.frontend.authentication.AuthenticationEngine;
 import org.apache.shardingsphere.proxy.frontend.authentication.AuthenticationResult;
@@ -50,22 +54,25 @@ public final class OpenGaussAuthenticationEngine implements AuthenticationEngine
     
     private static final int SSL_REQUEST_CODE = 80877103;
     
-    private boolean startupMessageReceived;
+    private static final int PROTOCOL_351_SERVER_ITERATOR = 10000;
     
-    private String clientEncoding;
+    private static final int PROTOCOL_350_SERVER_ITERATOR = 2048;
     
     private final String saltHexString;
     
     private final String nonceHexString;
     
-    private final int serverIteration;
+    private boolean startupMessageReceived;
+    
+    private String clientEncoding;
+    
+    private int serverIteration;
     
     private AuthenticationResult currentAuthResult;
     
     public OpenGaussAuthenticationEngine() {
         saltHexString = generateRandomHexString(64);
         nonceHexString = generateRandomHexString(8);
-        serverIteration = 10000;
     }
     
     private String generateRandomHexString(final int length) {
@@ -101,9 +108,21 @@ public final class OpenGaussAuthenticationEngine implements AuthenticationEngine
         if (Strings.isNullOrEmpty(user)) {
             throw new EmptyUsernameException();
         }
-        context.writeAndFlush(new OpenGaussAuthenticationSCRAMSha256Packet(saltHexString.getBytes(), nonceHexString.getBytes(), serverIteration));
+        serverIteration = comStartupPacket.getVersion() == OpenGaussProtocolVersion.PROTOCOL_351.getVersion() ? PROTOCOL_351_SERVER_ITERATOR : PROTOCOL_350_SERVER_ITERATOR;
+        String serverSignature = calculateServerSignature(comStartupPacket.getVersion(), user);
+        context.writeAndFlush(new OpenGaussAuthenticationSCRAMSha256Packet(
+                comStartupPacket.getVersion(), saltHexString.getBytes(), nonceHexString.getBytes(), serverSignature.getBytes(), serverIteration));
         currentAuthResult = AuthenticationResultBuilder.continued(user, "", comStartupPacket.getDatabase());
         return currentAuthResult;
+    }
+    
+    private String calculateServerSignature(final int version, final String username) {
+        if (version >= OpenGaussProtocolVersion.PROTOCOL_350.getVersion()) {
+            return "";
+        }
+        return ProxyContext.getInstance().getContextManager().getMetaDataContexts().getMetaData().getGlobalRuleMetaData().findSingleRule(AuthorityRule.class)
+                .flatMap(authorityRule -> authorityRule.findUser(new Grantee(username, "%")))
+                .map(user -> OpenGaussAuthenticationHandler.calculateServerSignature(user.getPassword(), saltHexString, nonceHexString, serverIteration)).orElse("");
     }
     
     private AuthenticationResult processPasswordMessage(final ChannelHandlerContext context, final PostgreSQLPacketPayload payload) {

--- a/proxy/frontend/opengauss/src/main/java/org/apache/shardingsphere/proxy/frontend/opengauss/authentication/OpenGaussAuthenticationEngine.java
+++ b/proxy/frontend/opengauss/src/main/java/org/apache/shardingsphere/proxy/frontend/opengauss/authentication/OpenGaussAuthenticationEngine.java
@@ -36,6 +36,7 @@ import org.apache.shardingsphere.db.protocol.postgresql.payload.PostgreSQLPacket
 import org.apache.shardingsphere.dialect.postgresql.exception.authority.EmptyUsernameException;
 import org.apache.shardingsphere.dialect.postgresql.exception.protocol.ProtocolViolationException;
 import org.apache.shardingsphere.infra.metadata.user.Grantee;
+import org.apache.shardingsphere.infra.metadata.user.ShardingSphereUser;
 import org.apache.shardingsphere.proxy.backend.context.ProxyContext;
 import org.apache.shardingsphere.proxy.backend.handler.admin.postgresql.PostgreSQLCharacterSets;
 import org.apache.shardingsphere.proxy.frontend.authentication.AuthenticationEngine;
@@ -120,9 +121,9 @@ public final class OpenGaussAuthenticationEngine implements AuthenticationEngine
         if (version >= OpenGaussProtocolVersion.PROTOCOL_350.getVersion()) {
             return "";
         }
-        return ProxyContext.getInstance().getContextManager().getMetaDataContexts().getMetaData().getGlobalRuleMetaData().findSingleRule(AuthorityRule.class)
-                .flatMap(authorityRule -> authorityRule.findUser(new Grantee(username, "%")))
-                .map(user -> OpenGaussAuthenticationHandler.calculateServerSignature(user.getPassword(), saltHexString, nonceHexString, serverIteration)).orElse("");
+        String password = ProxyContext.getInstance().getContextManager().getMetaDataContexts().getMetaData().getGlobalRuleMetaData().findSingleRule(AuthorityRule.class)
+                .flatMap(authorityRule -> authorityRule.findUser(new Grantee(username, "%"))).map(ShardingSphereUser::getPassword).orElse("");
+        return OpenGaussAuthenticationHandler.calculateServerSignature(password, saltHexString, nonceHexString, serverIteration);
     }
     
     private AuthenticationResult processPasswordMessage(final ChannelHandlerContext context, final PostgreSQLPacketPayload payload) {

--- a/proxy/frontend/opengauss/src/main/java/org/apache/shardingsphere/proxy/frontend/opengauss/authentication/OpenGaussAuthenticationHandler.java
+++ b/proxy/frontend/opengauss/src/main/java/org/apache/shardingsphere/proxy/frontend/opengauss/authentication/OpenGaussAuthenticationHandler.java
@@ -37,6 +37,7 @@ import javax.crypto.Mac;
 import javax.crypto.SecretKeyFactory;
 import javax.crypto.spec.PBEKeySpec;
 import javax.crypto.spec.SecretKeySpec;
+import java.nio.charset.StandardCharsets;
 import java.security.InvalidKeyException;
 import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
@@ -60,7 +61,25 @@ public final class OpenGaussAuthenticationHandler {
     
     private static final String SHA256_ALGORITHM = "SHA-256";
     
+    private static final String SERVER_KEY = "Server Key";
+    
     private static final String CLIENT_KEY = "Client Key";
+    
+    /**
+     * Calculate server signature.
+     *
+     * @param password password
+     * @param salt salt in hex string
+     * @param nonce nonce in hex string
+     * @param serverIteration server iteration
+     * @return server signature
+     */
+    public static String calculateServerSignature(final String password, final String salt, final String nonce, final int serverIteration) {
+        byte[] k = generateKFromPBKDF2(password, salt, serverIteration);
+        byte[] serverKey = getKeyFromHmac(k, SERVER_KEY.getBytes(StandardCharsets.UTF_8));
+        byte[] result = getKeyFromHmac(serverKey, hexStringToBytes(nonce));
+        return bytesToHexString(result);
+    }
     
     /**
      * Login with SCRAM SHA-256 password.
@@ -146,6 +165,18 @@ public final class OpenGaussAuthenticationHandler {
             result[i] = (byte) (charToByte(hexChars[pos]) << 4 | charToByte(hexChars[pos + 1]));
         }
         return result;
+    }
+    
+    private static String bytesToHexString(byte[] src) {
+        StringBuilder result = new StringBuilder();
+        for (byte each : src) {
+            String hex = Integer.toHexString(each & 255);
+            if (hex.length() < 2) {
+                result.append(0);
+            }
+            result.append(hex);
+        }
+        return result.toString();
     }
     
     private static byte charToByte(final char c) {

--- a/proxy/frontend/opengauss/src/main/java/org/apache/shardingsphere/proxy/frontend/opengauss/authentication/OpenGaussAuthenticationHandler.java
+++ b/proxy/frontend/opengauss/src/main/java/org/apache/shardingsphere/proxy/frontend/opengauss/authentication/OpenGaussAuthenticationHandler.java
@@ -81,6 +81,18 @@ public final class OpenGaussAuthenticationHandler {
         return bytesToHexString(result);
     }
     
+    private static String bytesToHexString(final byte[] src) {
+        StringBuilder result = new StringBuilder();
+        for (byte each : src) {
+            String hex = Integer.toHexString(each & 255);
+            if (hex.length() < 2) {
+                result.append(0);
+            }
+            result.append(hex);
+        }
+        return result.toString();
+    }
+    
     /**
      * Login with SCRAM SHA-256 password.
      *
@@ -165,18 +177,6 @@ public final class OpenGaussAuthenticationHandler {
             result[i] = (byte) (charToByte(hexChars[pos]) << 4 | charToByte(hexChars[pos + 1]));
         }
         return result;
-    }
-    
-    private static String bytesToHexString(byte[] src) {
-        StringBuilder result = new StringBuilder();
-        for (byte each : src) {
-            String hex = Integer.toHexString(each & 255);
-            if (hex.length() < 2) {
-                result.append(0);
-            }
-            result.append(hex);
-        }
-        return result.toString();
     }
     
     private static byte charToByte(final char c) {

--- a/proxy/frontend/opengauss/src/test/java/org/apache/shardingsphere/proxy/frontend/opengauss/authentication/OpenGaussAuthenticationEngineTest.java
+++ b/proxy/frontend/opengauss/src/test/java/org/apache/shardingsphere/proxy/frontend/opengauss/authentication/OpenGaussAuthenticationEngineTest.java
@@ -133,6 +133,10 @@ public final class OpenGaussAuthenticationEngineTest extends ProxyContextRestore
     }
     
     private void assertLogin(final String inputPassword) {
+        MetaDataContexts metaDataContexts = getMetaDataContexts(new ShardingSphereUser(username, password, ""));
+        ContextManager contextManager = mock(ContextManager.class, RETURNS_DEEP_STUBS);
+        when(contextManager.getMetaDataContexts()).thenReturn(metaDataContexts);
+        ProxyContext.init(contextManager);
         PostgreSQLPacketPayload payload = new PostgreSQLPacketPayload(createByteBuf(16, 128), StandardCharsets.UTF_8);
         payload.writeInt4(64);
         payload.writeInt4(196608);
@@ -155,10 +159,6 @@ public final class OpenGaussAuthenticationEngineTest extends ProxyContextRestore
         payload.writeInt1('p');
         payload.writeInt4(4 + clientDigest.length() + 1);
         payload.writeStringNul(clientDigest);
-        MetaDataContexts metaDataContexts = getMetaDataContexts(new ShardingSphereUser(username, password, ""));
-        ContextManager contextManager = mock(ContextManager.class, RETURNS_DEEP_STUBS);
-        when(contextManager.getMetaDataContexts()).thenReturn(metaDataContexts);
-        ProxyContext.init(contextManager);
         actual = engine.authenticate(channelHandlerContext, payload);
         assertThat(actual.isFinished(), is(password.equals(inputPassword)));
     }

